### PR TITLE
Bridge

### PR DESCRIPTION
**Pull request by ghost**
Cluster definitions are automatically generated from the IDL and
devices automatically route read/writes to the appropriate cluster.

Writes can be intercepted on a per-cluster basis to talk to whatever
bridged backend there is.

Bridge includes an interactive CLI to modify and inspect bridge state

b/225921621

Testing
Manual testing with chip-tool on linux.
Add device to bridge via CLI
new-device switch
finish-device

Then read the newly added switch device's attributes
Then change attribute via set and read new value

### CODE REVIEW

File: examples/bridge-app/linux/include/Device.h [hunk 1]
```diff
@@ -18,230 +18,59 @@
...
-    inline void SetEndpointId(chip::EndpointId id) { mEndpointId = id; };
-    inline chip::EndpointId GetEndpointId() { return mEndpointId; };
-    inline void SetParentEndpointId(chip::EndpointId id) { mParentEndpointId = id; };
-    inline chip::EndpointId GetParentEndpointId() { return mParentEndpointId; };
-    inline char * GetName() { return mName; };
-    inline std::string GetLocation() { return mLocation; };
-    inline std::string GetZone() { return mZone; };
-    inline void SetZone(std::string zone) { mZone = zone; };
+    const char * GetName();
+    void SetName(const char * name);
```
Comment (andy31415) [inline examples/bridge-app/linux/include/Device.h:63]: Should we have a comment regarding lifetimes for this?

or given that we are on a large platform, how about storing the name as a std::string and not worry about lifetimes of pointers.

File: scripts/idl/BUILD.gn [hunk 2]
```diff
@@ -25,6 +25,7 @@ pw_python_package("idl") {
     "matter_grammar.lark",
 
     # Templates used for generation
+    "generators/cpp/BridgeClustersCpp.jinja",
```
Comment (andy31415) [inline scripts/idl/BUILD.gn:28]: Combining a generic `cpp` name for generators and a template that is `BridgeClustersCpp` seems mismatched.

are these cpp generic (in which case the jinja template should have a different name) or are they bridge specific (in which case the generator name should be named differently than cpp to make it clear what it generates.

File: scripts/idl/generators/cpp/BridgeClustersCpp.jinja [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+    return t.Encode(writer, tag);
+  }
+
+  {%-   for field in struct.fields %}
+  {{field | getField(None, idl)}} {{field.name}};
+  {%-  endfor  %}
+};
+{%- endfor %}
+
+{%- for cluster in clusters %}
```
Comment (andy31415) [inline scripts/idl/generators/cpp/BridgeClustersCpp.jinja:37]: Can we split files into per-cluster or similar? 

ZAP "generate single file" is very painful in compile times and debuggability/inclusion. Could we generate one cluster struct (and potentially encode/decode) for each cluster? I would rather do things like `#include <LightBridge>` than `#include <AllBridges>`

File: scripts/idl/generators/cpp/BridgeClustersCpp.jinja [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+{%-   if cluster | dynamicCluster(idl) %}
+struct {{cluster.name}}Cluster : public CommonCluster
```
Comment (andy31415) [inline scripts/idl/generators/cpp/BridgeClustersCpp.jinja:39]: Is this generating a header  only? Could we split generating a header and cpp file?

File: scripts/idl/tests/outputs/cluster_struct_attribute/cpp/BridgeClustersImpl.h [hunk 5]
```diff
@@ -0,0 +1,148 @@
+#include <app-common/zap-generated/cluster-id.h>
+#include <app-common/zap-generated/attribute-id.h>
+
+#include <new>
+
+namespace clusters {
+
+struct DemoClusterCluster : public CommonCluster
```
Comment (andy31415) [inline scripts/idl/tests/outputs/cluster_struct_attribute/cpp/BridgeClustersImpl.h:8]: I think we should split the cluster struct from clusterinfo and attrinfo.

I believe that code would be easier to maintain if the generated code looks somewhat similar to what one would manually generate (minus constants or some formatting).

File: scripts/idl/generators/cpp/BridgeClustersCpp.jinja (hunk 4 above)
Comment (achaulk-goog) [reply scripts/idl/generators/cpp/BridgeClustersCpp.jinja:39]: That gets tricky as the generated code depends on classes declared by the bridge app, so moving things around in the bridge app could mean changing the codegen to output new include paths. If it stays header-only then the only requirement is that these structs be declared somewhere prior to inclusion

Comment (andy31415) [reply scripts/idl/generators/cpp/BridgeClustersCpp.jinja:39]: Could you explain with an example? trying to wrap my head around the statement of 'depends on classes declared by bridge app'. Is this a circular include case?

Comment (achaulk-goog) [reply scripts/idl/generators/cpp/BridgeClustersCpp.jinja:39]: Not circular includes no, but the Attribute template is defined in bridge-app/linux/include/Clusters.h and is used in the generated code, so any generated file either has to include that, or get included after that header

File: scripts/idl/generators/cpp/BridgeClustersCpp.jinja (hunk 3 above)
Comment (achaulk-goog) [reply scripts/idl/generators/cpp/BridgeClustersCpp.jinja]: There's still a single include point the generated cluster & attribute list does the including but it's split out now

File: examples/bridge-app/linux/include/Clusters.h [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+    using Type::operator=;
+};
+
+struct CommonCluster;
+typedef std::function<CHIP_ERROR(CommonCluster *, const chip::app::ConcreteDataAttributePath &, chip::app::AttributeValueDecoder &)>
+    PropagateWriteCB;
+
+// This is the base type of all generated clusters, providing the backend access to
+// implementation details.
+struct CommonCluster : public ClusterImpl
```
Comment (mrjerryjohns) [inline examples/bridge-app/linux/include/Clusters.h:223]: The namings here are a bit odd - `ClusterImpl` is an interface, but it contains the term 'Impl' in it.

Recommendation: Rename to `CommonClusterImpl` and `ClusterInterface`.

File: examples/bridge-app/linux/include/Clusters.h [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+static constexpr uint16_t ZCL_DESCRIPTOR_CLUSTER_REVISION              = 1;
+static constexpr uint16_t ZCL_BRIDGED_DEVICE_BASIC_CLUSTER_REVISION    = 1;
+static constexpr uint16_t ZCL_FIXED_LABEL_CLUSTER_REVISION             = 1;
+static constexpr uint16_t ZCL_ON_OFF_CLUSTER_REVISION                  = 4;
+static constexpr uint16_t ZCL_SWITCH_CLUSTER_REVISION                  = 1;
+static constexpr uint16_t ZCL_LEVEL_CONTROL_CLUSTER_REVISION           = 1;
+static constexpr uint16_t ZCL_TEMPERATURE_MEASUREMENT_CLUSTER_REVISION = 1;
+
+// This is the interface to cluster implementations, providing access to manipulate attributes.
+class ClusterImpl
```
Comment (mrjerryjohns) [inline examples/bridge-app/linux/include/Clusters.h:40]: This should be separated out into its own header file.

File: examples/bridge-app/linux/include/Clusters.h [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+#include <app/ConcreteAttributePath.h>
+#include <app/reporting/reporting.h>
+#include <app/util/attribute-storage.h>
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include <functional>
+
+static constexpr uint16_t ZCL_DESCRIPTOR_CLUSTER_REVISION              = 1;
```
Comment (mrjerryjohns) [inline examples/bridge-app/linux/include/Clusters.h:31]: Why are these here?

File: examples/bridge-app/linux/include/Clusters.h [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+static constexpr uint16_t ZCL_LEVEL_CONTROL_CLUSTER_REVISION           = 1;
+static constexpr uint16_t ZCL_TEMPERATURE_MEASUREMENT_CLUSTER_REVISION = 1;
+
+// This is the interface to cluster implementations, providing access to manipulate attributes.
+class ClusterImpl
+{
+public:
+    virtual ~ClusterImpl() = default;
+
+    virtual void SetEndpointId(chip::EndpointId id)                          = 0;
```
Comment (mrjerryjohns) [inline examples/bridge-app/linux/include/Clusters.h:45]: Function documentation is missing here, and everywhere else.

File: examples/bridge-app/linux/include/Clusters.h [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+    T Peek() const { return mValue; }
+
+private:
+    T mValue;
+    static_assert(std::is_standard_layout<T>::value, "PrimitiveType not standard layout!");
+};
+
+// This provides access to a struct. Each struct generates its own access logic.
+template <typename T, bool IsFabricScoped = false>
+struct StructType : public T
```
Comment (mrjerryjohns) [inline examples/bridge-app/linux/include/Clusters.h:90]: Given my comment in `AddAllAttributes`, I don't think these extra type definitions are warranted, since without the `kMatterType` field, the rest of it is really just a shallow wrapper around the existing templated `AttributeValueDecoder::Decode` and `AttributeValueEncoder::Encode` functions. 

I'd rather we add more types to the set of types that can be passed in there, rather than wrapping them. That would allow you to avoid defining `StructType` and `PrimitiveType` completely.

As for the variable-collection types like octet string and lists, we should create new types backed by `std::string` and `std:::vector` that are usable by `DataModel::Encode` and `DataModel::Decode`. That will be way way more scalable option, and usable by others than just bridge apps.

**Review by mrjerryjohns — CHANGES_REQUESTED** [review]

File: examples/bridge-app/linux/include/Clusters.h (hunk 8 above)
Comment (achaulk-goog) [reply examples/bridge-app/linux/include/Clusters.h:31]: Default initialization of the cluster revision attribute

File: examples/bridge-app/linux/include/Clusters.h [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+
+    chip::Span<const EmberAfAttributeMetadata> GetAllAttributes() override
+    {
+        static constexpr EmberAfAttributeMetadata kAllAttributes[] = {
+            { ZCL_REACHABLE_ATTRIBUTE_ID, ZCL_BOOLEAN_ATTRIBUTE_TYPE, 1, 0, 0u },
+        };
+        return chip::Span<const EmberAfAttributeMetadata>(kAllAttributes);
+    }
+
+    Attribute<ZCL_REACHABLE_ATTRIBUTE_ID, 0, PrimitiveType<uint8_t, 1, ZCL_BOOLEAN_ATTRIBUTE_TYPE>> mReachable;
```
[the remark this answers is not in the document]
Comment (achaulk-goog) [reply examples/bridge-app/linux/include/Clusters.h:291]: Add it or not add it to the list of attributes for that cluster. Presumably any access to an attribute not in that list is already an error so doing dummy accesses should be acceptable. Otherwise we'll have to store a list of enabled attributes in here

File: examples/bridge-app/linux/include/Clusters.h [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+private:
+    uint16_t mLength = 0xFFFF;
+    T mArray[MaxElements];
+};
+
+// This is used by code generation to provide specific attributes. Type should be one of
+// PrimitiveType, OctetString, ArrayType, etc to provide storage and type-specific
+// functionality.
+template <chip::AttributeId id, EmberAfAttributeMask mask, typename Type>
+struct Attribute : public Type
```
[the remark this answers is not in the document]
Comment (achaulk-goog) [reply examples/bridge-app/linux/include/Clusters.h:208]: Optional would be the responsibility of an upper layer to filter the requests. Nullable is implemented by Type, though it does look like I missed it for primitives and structs

File: examples/bridge-app/linux/include/Clusters.h [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+    static constexpr chip::ClusterId kClusterId = ZCL_BRIDGED_DEVICE_BASIC_CLUSTER_ID;
+
+    BridgedDeviceBasicCluster();
+
+    chip::ClusterId GetClusterId() override { return kClusterId; }
+    CHIP_ERROR WriteFromBridge(const chip::app::ConcreteDataAttributePath & aPath,
+                               chip::app::AttributeValueDecoder & aDecoder) override;
+
+    template <typename T>
+    void AddAllAttributes(T * list)
```
[the remark this answers is not in the document]
Comment (achaulk-goog) [reply examples/bridge-app/linux/include/Clusters.h:278]: Yeah this is mostly redundant now, I'll remove

File: examples/bridge-app/linux/include/Clusters.h [hunk 14]
```diff
@@ -0,0 +1,294 @@
...
+struct CommonAttributeAccessInterface : public chip::app::AttributeAccessInterface
+{
+    using chip::app::AttributeAccessInterface::AttributeAccessInterface;
+    static CommonCluster * FindCluster(const chip::app::ConcreteClusterPath & path);
+};
+
+#include "bridge/BridgeClustersImpl.h"
+
+namespace clusters {
+struct BridgedDeviceBasicCluster : public CommonCluster
```
[the remark this answers is not in the document]
Comment (achaulk-goog) [reply examples/bridge-app/linux/include/Clusters.h:267]: This is required but missing from the IDL, seemed easier to just add here for now

File: examples/bridge-app/linux/include/Clusters.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (mrjerryjohns) [reply examples/bridge-app/linux/include/Clusters.h:291]: But this is code-generated right? How would an application developer using this code-generated class be able to impart that level of customization?

File: examples/bridge-app/linux/include/Clusters.h (hunk 8 above)
Comment (mrjerryjohns) [reply examples/bridge-app/linux/include/Clusters.h:31]: This file is too much a mash-up of library APIs, and cluster-specific bits. Please separate.

File: examples/bridge-app/linux/include/Clusters.h (hunk 10 above)
Comment (achaulk-goog) [reply examples/bridge-app/linux/include/Clusters.h:90]: Lists can't be easily handled at the DataModel level because you're missing the context of the ListOperation and/or index

I've got a version where strings are specialized at the DM layer and lists are specialized at the top level and everything else is merged

File: examples/bridge-app/linux/include/Clusters.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (achaulk-goog) [reply examples/bridge-app/linux/include/Clusters.h]: Yes, the default list adds all attributes, but if a specific app wanted to support fewer attributes then it could when building the device, adding only the attributes it supports. Different devices can support different sets of attributes for the same cluster type as well. Space would still be allocated for every attribute which is slightly wasteful, but it would work. Alternatively, the IDL could be edited to remove unsupported attributes entirely

File: examples/bridge-app/linux/include/Device.h (hunk 1 above)
Comment (achaulk-goog) [reply examples/bridge-app/linux/include/Device.h:63]: Just made it std::string. Can worry about lifetimes later if needed

**Comment by mrjerryjohns** [comment]
As per our conversation, here're the four things we should fix to make this more re-usable and integratable into the broader SDK in the near future, since I think there are a number of concepts here worth leveraging going forward:

1. Nullable and Optional specialization of `Attribute<T>` is inverted with what we have in the SDK's cluster object types. I'd rather we flip it to make it similar, i.e

`Nullable<ArrayAttribute<int>>`, or `Nullable<ArrayAttribute>`, etc.

To do that, we might need to add a new file similar to `src/app/data-model/Encode.h, Decode.h` containing methods that permit path to provided to these 'container', non-view classes:

```
template <typename X, std::enable_if_t<DataModel::IsList<X>::value, bool> = true>
CHIP_ERROR EncodeWithPath(ConcreteDataAttributePath &path, chip::app::AttributeValueEncoder & encoder, X &x)
{
     return x.Encode(path, encoder);
}

template <typename X, std::enable_if_t<DataModel::IsList<X>::value, bool> = false>     
CHIP_ERROR EncodeWithPath(ConcreteDataAttributePath &path, chip::app::AttributeValueEncoder & encoder, X &x)
{
     return encoder.Encode(x);
}
```

2. Create an `AttributeInterface` class that has pure virtuals for:

    CHIP_ERROR Write(const chip::app::ConcreteDataAttributePath & aPath, chip::app::AttributeValueDecoder & aDecoder) = 0
    CHIP_ERROR Read(const chip::app::ConcreteReadAttributePath & aPath, chip::app::AttributeValueEncoder & aEncoder) = 0

As well as containing the metadata information. This then makes it easy to track the attributes in a list, as well as dynamically create the metadata table from this very information.

3. `ClusterInterface` could then pivot to having a list of `AttributeInterface` items.
4. To simplify the PR, remove the fixed, non-stl variants and just have the STL variants for now for bridges. This minimizes the scope of the PR and makes it more easy to integrate.

File: examples/dynamic-bridge-app/linux/include/DataModel.h [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+
+
+/*
+ * @brief
+ *
+ * If an item is requested from a list, encode just that single item, or the entire list otherwise.
+ *
+ * The object must satisfy the following constraints
+ * size() must return an integer
+ * begin() must return a type conforming to LegacyRandomAccessIterator
```
Comment (mrjerryjohns) [inline examples/dynamic-bridge-app/linux/include/DataModel.h:60]: Does it have to be a `LegacyRandomAccessIterator`? Couldn't we relax this to be a `LegacyForwardIterator` to then permit use of `std::list` (and other similar containers) if needed as well?

File: examples/dynamic-bridge-app/linux/include/DataModel.h [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+
+
+template<typename X, std::enable_if_t<IsList<std::decay_t<X>>::value && (sizeof(std::decay_t<typename X::pointer>) > sizeof(char)), bool> = true>
+CHIP_ERROR Decode(const ConcreteDataAttributePath& aPath, AttributeValueDecoder & aDecoder, X& x)
+{
+    switch(aPath.mListOp) {
+    case ConcreteDataAttributePath::ListOperation::DeleteItem:
+        if (aPath.mListIndex >= x.size())
+            return CHIP_ERROR_INVALID_LIST_LENGTH;
+        x.erase(x.begin() + aPath.mListIndex);
```
Comment (mrjerryjohns) [inline examples/dynamic-bridge-app/linux/include/DataModel.h:166]: The expectations of the container and the methods it should have should be clearly defined at the top of the file for  sequence-type containers.

File: examples/dynamic-bridge-app/linux/include/DataModel.h [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+        x = X(span.data(), span.size());
+    }
+    return err;
+}
+
+
+template<typename X, std::enable_if_t<IsList<std::decay_t<X>>::value && (sizeof(std::decay_t<typename X::pointer>) > sizeof(char)), bool> = true>
+CHIP_ERROR Decode(const ConcreteDataAttributePath& aPath, AttributeValueDecoder & aDecoder, X& x)
+{
+    switch(aPath.mListOp) {
```
Comment (mrjerryjohns) [inline examples/dynamic-bridge-app/linux/include/DataModel.h:162]: What about the list ReplaceAll operation?

File: examples/dynamic-bridge-app/linux/include/DataModel.h [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+        x.erase(x.begin() + aPath.mListIndex);
+        return CHIP_NO_ERROR;
+
+    case ConcreteDataAttributePath::ListOperation::ReplaceItem:
+        if (aPath.mListIndex >= x.size())
+            return CHIP_ERROR_INVALID_LIST_LENGTH;
+        return aDecoder.Decode(x.begin()[aPath.mListIndex]);
+
+    default:
+        x.emplace_back();
```
Comment (mrjerryjohns) [inline examples/dynamic-bridge-app/linux/include/DataModel.h:175]: Why not construct the element type of X temporarily on stack, and then call push_back? That would presumably make this work with lists then?

File: examples/dynamic-bridge-app/linux/include/Attribute.h [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+    Attribute(chip::CharSpan name, chip::AttributeId id, EmberAfAttributeMask mask, EmberAfAttributeType type, size_t size, Type value = Type())
+    	: mMetadata(EmberAfAttributeMetadata{id, type, (uint16_t)size, mask, ZAP_EMPTY_DEFAULT()}), mData(value), mName(name)
+    {
+    }
+
+    chip::AttributeId GetId() override { return mMetadata.attributeId; }
+    const EmberAfAttributeMetadata& GetMetadata() override { return mMetadata; }
+    chip::CharSpan GetName() override { return mName; }
+
+    CHIP_ERROR Write(const chip::app::ConcreteDataAttributePath & aPath, chip::app::AttributeValueDecoder & aDecoder) override
```
Comment (mrjerryjohns) [inline examples/dynamic-bridge-app/linux/include/Attribute.h:44]: These two methods now look much more regular/normalized. Awesome!

File: examples/dynamic-bridge-app/linux/include/Attribute.h [hunk 20]
```diff
@@ -0,0 +1,80 @@
+#pragma once
+
+#include <app/ConcreteAttributePath.h>
+
+#include "DataModel.h"
+
+// This is the interface to the attribute implementation
```
Comment (mrjerryjohns) [inline examples/dynamic-bridge-app/linux/include/Attribute.h:7]: "...that permits retrieval of metadata about the attribute as well as set/get the underlying value"

File: examples/dynamic-bridge-app/linux/include/Clusters.h [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+    virtual chip::ClusterId GetClusterId() = 0;
+    // Gets the list of enabled attributes for this cluster.
+    virtual chip::Span<AttributeInterface *> GetAllAttributes() = 0;
+    // Find an attribute of this cluster, if present.
+    virtual AttributeInterface * FindAttribute(chip::AttributeId id) = 0;
+    // Find an attribute of this cluster, if present, given its name.
+    virtual AttributeInterface * FindAttributeByName(chip::CharSpan name) = 0;
+    // Write the given TLV encoded value to the given attribute. Can write
+    // normally read-only values. Nofitied matter that the value has changed.
+    virtual bool Push(chip::AttributeId attr, chip::TLV::TLVReader & reader) = 0;
```
Comment (mrjerryjohns) [inline examples/dynamic-bridge-app/linux/include/Clusters.h:52]: The nomenclature so far has been "Write" and "Read". Why not continue with that consistent naming convention?

File: examples/dynamic-bridge-app/linux/include/Clusters.h [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+    chip::EndpointId GetEndpointId() const;
+
+    chip::ClusterId GetClusterId() override;
+
+    void SetCallback(PropagateWriteCB * cb);
+
+    bool active() const;
+
+    // Implemented by generated code. Valid writes should be forwarded to ForwardWriteToBridge.
+    virtual CHIP_ERROR WriteFromBridge(const chip::app::ConcreteDataAttributePath & aPath,
```
Comment (mrjerryjohns) [inline examples/dynamic-bridge-app/linux/include/Clusters.h:77]: Any reason this is marked virtual? The default impl provided here should be sufficient for all derived classes no?

File: examples/dynamic-bridge-app/linux/include/Device.h [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+#define DEVICE_TYPE_LO_ON_OFF_LIGHT 0x0100
+// (taken from lo-devices.xml)
+#define DEVICE_TYPE_LO_ON_OFF_LIGHT_SWITCH 0x0103
+// (taken from chip-devices.xml)
+#define DEVICE_TYPE_ROOT_NODE 0x0016
+// (taken from chip-devices.xml)
+#define DEVICE_TYPE_BRIDGE 0x000e
+
+// Device Version for dynamic endpoints:
+#define DEVICE_VERSION_DEFAULT 1
```
Comment (mrjerryjohns) [inline examples/dynamic-bridge-app/linux/include/Device.h:42]: Please separate out these into their own 'DeviceTypesList.h` or something.

File: examples/dynamic-bridge-app/linux/include/Device.h [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+    chip::EndpointId mEndpointId;
+    chip::Span<chip::DataVersion> mDataVersions;
+    chip::Span<EmberAfCluster> mClusters;
+    chip::Span<ClusterInterface *> mClusterImpl;
+    chip::Span<const EmberAfDeviceType> mDeviceTypeList;
+    EmberAfEndpointType mEndpointType;
+    std::string mDeviceName;
+};
+
+class Action
```
Comment (mrjerryjohns) [inline examples/dynamic-bridge-app/linux/include/Device.h:76]: This belongs in its own `ActionCluster.h` file.

File: examples/dynamic-bridge-app/linux/include/GeneratedClusters.h [hunk 25]
```diff
@@ -0,0 +1,77 @@
...
+#include <stdbool.h>
+#include <stdint.h>
+
+#include <functional>
+
+#include "Attribute.h"
+#include "Clusters.h"
+#include "DataModel.h"
+
+struct CommonAttributeAccessInterface : public chip::app::AttributeAccessInterface
```
Comment (mrjerryjohns) [inline examples/dynamic-bridge-app/linux/include/GeneratedClusters.h:17]: Can we please separate this out from the cluster-specific bits in this file?

File: examples/dynamic-bridge-app/linux/include/Attribute.h [hunk 26]
```diff
@@ -0,0 +1,80 @@
+#pragma once
```
Comment (mrjerryjohns) [inline examples/dynamic-bridge-app/linux/include/Attribute.h:1]: To better organize this stuff, can we please create a 'data-model' sub-folder which contains the *.cpp and *.h files for the common, library bits, and a separate 'generated' folder that contains all the clusters and device type bits?

File: examples/dynamic-bridge-app/linux/include/DataModel.h (hunk 17 above)
Comment (achaulk-goog) [reply examples/dynamic-bridge-app/linux/include/DataModel.h:162]: Not sure it's possible to do here without extra state since we would just get decode, decode, decode, decode and none of them would know here if it's the first one. Unless that's set for only one?

File: examples/dynamic-bridge-app/linux/include/DataModel.h [hunk 27]
```diff
@@ -0,0 +1,218 @@
...
+    {
+        uint16_t index = aPath.mListIndex.Value();
+        if (index >= x.size())
+            return CHIP_ERROR_INVALID_ARGUMENT;
+
+        return aEncoder.Encode(x.data()[index]);
+    }
+    return aEncoder.EncodeList([x](const auto & encoder) {
+        CHIP_ERROR err = CHIP_NO_ERROR;
+        for(auto& v : x) {
```
[the remark this answers is not in the document]
Comment (achaulk-goog) [reply examples/dynamic-bridge-app/linux/include/DataModel.h:78]: Yes, anything that supports begin(), end(), ++iterator and *iterator supports ranged for

File: examples/dynamic-bridge-app/linux/include/DataModel.h (hunk 15 above)
Comment (achaulk-goog) [reply examples/dynamic-bridge-app/linux/include/DataModel.h:60]: Yes if we are willing to repeatedly increment the iterator instead of assuming we can use indexing

File: examples/dynamic-bridge-app/linux/include/DataModel.h (hunk 18 above)
Comment (achaulk-goog) [reply examples/dynamic-bridge-app/linux/include/DataModel.h:175]: Are there things that don't support emplace_back? std::list does

File: examples/dynamic-bridge-app/linux/include/Clusters.h (hunk 21 above)
Comment (achaulk-goog) [reply examples/dynamic-bridge-app/linux/include/Clusters.h:52]: Read/Write are for matter side requests, this is logically a bridge-side out-of-band change, so write + notify matter that it has changed

File: examples/dynamic-bridge-app/linux/include/Clusters.h (hunk 22 above)
Comment (achaulk-goog) [reply examples/dynamic-bridge-app/linux/include/Clusters.h:77]: Probably not with the new type structure no

File: scripts/idl/generators/cpp/BridgeClustersCpp.jinja (hunk 4 above)
Comment (mrjerryjohns) [reply scripts/idl/generators/cpp/BridgeClustersCpp.jinja]: The generated code should be including the `bridge-app/linux/include/Clusters.h`.

File: examples/bridge-app/linux/include/Clusters.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (mrjerryjohns) [reply examples/bridge-app/linux/include/Clusters.h]: Resolved as per new implementation.

File: examples/bridge-app/linux/include/Clusters.h (hunk 10 above)
Comment (mrjerryjohns) [reply examples/bridge-app/linux/include/Clusters.h]: This has now been updated.

File: examples/dynamic-bridge-app/linux/include/DataModel.h (hunk 17 above)
Comment (mrjerryjohns) [reply examples/dynamic-bridge-app/linux/include/DataModel.h:162]: I'm talking about dealing with `ConcreteDataAttributePath::ListOperation::ReplaceAll`.

File: examples/dynamic-bridge-app/linux/include/DataModel.h (hunk 18 above)
Comment (mrjerryjohns) [reply examples/dynamic-bridge-app/linux/include/DataModel.h:175]: Ah nvm. You're right.

File: examples/dynamic-bridge-app/linux/include/Clusters.h (hunk 21 above)
Comment (mrjerryjohns) [reply examples/dynamic-bridge-app/linux/include/Clusters.h:52]: That's imbuing some notion of how this API is utilized by consumers into its naming, which seems un-necessary I think.

File: examples/dynamic-bridge-app/linux/include/Device.h (hunk 23 above)
Comment (achaulk-goog) [reply examples/dynamic-bridge-app/linux/include/Device.h:42]: Moved the only used ones to cpp

File: examples/dynamic-bridge-app/linux/include/DataModel.h (hunk 17 above)
Comment (achaulk-goog) [reply examples/dynamic-bridge-app/linux/include/DataModel.h:162]: It looks like it always replaces with a single element? That's doable

File: examples/dynamic-bridge-app/linux/include/DataModel.h (hunk 15 above)
Comment (achaulk-goog) [reply examples/dynamic-bridge-app/linux/include/DataModel.h:60]: Changed to use std::advance

**Review by mrjerryjohns — APPROVED** [review]
